worlds/witness/hints.py

```diff
@@ -338,22 +340,92 @@ def get_priority_hint_locations(world: "WitnessWorld") -> List[str]:
+
+    # Find valid location groups.
+    for group, locations in possible_location_groups.items():
+        if group == "Everywhere":
+            continue
+        present_locations = sum(location in locations_in_that_world for location in locations)
+        valid_location_groups[group] = present_locations
+
+    # If there are valid location groups, use a random one.
+    if valid_location_groups:
+        # If there are location groups with more than 1 location, remove any that only have 1.
+        if any(num_locs > 1 for num_locs in valid_location_groups.values()):
+            valid_location_groups = {name: num_locs for name, num_locs in valid_location_groups.items() if num_locs > 1}
+
+        location_groups_with_weights = {
+            # Listen. Just don't worry about it. :)))
+            location_group: (x ** 0.6) * math.e ** (- (x / 7) ** 0.6) if x > 6 else x / 6
+            for location_group, x in valid_location_groups.items()
+        }
+
+        location_groups = list(location_groups_with_weights.keys())
+        weights = list(location_groups_with_weights.values())
+
+        return world.random.choices(location_groups, weights, k=1)[0], "Group"
+
+    if allow_regions:
+        return hint_loc.parent_region.name, "Region"
+
+    return "Everywhere", "Everywhere"
+
+
 def word_direct_hint(world: "WitnessWorld", hint: WitnessLocationHint):
     location_name = hint.location.name
     if hint.location.player != world.player:
         location_name += " (" + world.multiworld.get_player_name(hint.location.player) + ")"
 
     item = hint.location.item
     item_name = item.name
+
     if item.player != world.player:
         item_name += " (" + world.multiworld.get_player_name(item.player) + ")"
 
-    if hint.hint_came_from_location:
-        hint_text = f"{location_name} contains {item_name}."
-    else:
-        hint_text = f"{item_name} can be found at {location_name}."
+    hint_text = ""
+    area: Optional[str] = None
+
+    if world.options.vague_hints:
+        chosen_group, group_type = try_getting_location_group_for_location(world, hint.location)
+
+        if hint.location.player == world.player:
+            area = chosen_group
 
-    return WitnessWordedHint(hint_text, hint.location)
+            # local locations should only ever return a location group, as Witness defines groups for every location.
+            hint_text = f"{item_name} can be found in the {area} area."
+        else:
+            player_name = world.multiworld.get_player_name(hint.location.player)
+
+            if group_type == "Everywhere":
+                location_name = f"a location in {player_name}'s world"
+            elif group_type == "Group":
+                location_name = f"a \"{chosen_group}\" location in {player_name}'s world"
+            elif group_type == "Region":
+                if chosen_group == "Menu":
+                    location_name = f"a location near the start of {player_name}'s game (\"Menu\" region)"
```

Comment:
This isn't _necessarily_ accurate. To again use Emerald as context, while the dexsanity locations are technically in a specific pokedex region for organizational reasons, I could have attached all dex entries directly to `Menu` for the same behavior. I actually wouldn't be surprised if the fact that a location was on `Menu` wasn't more indicative that it was special in some way (or that the implementation simply doesn't have regions), rather than in or near sphere 1.

I know all of Dark Souls III's consumable locations are attached to `Menu` because they're progressive locations with no logical requirements. `Titanite Chunk #20` is not a place you can go, it's the twentieth Titanite Chunk you pick up. And for that reason, it's attached directly to `Menu`. Yet a Dark Souls III player would probably not understand picking up 20 Titanite Chunks to be "near the start".

Reply:
This is just an extra comment to make sure this was considered. Whether any changes are made or not made as a result don't have a bearing on the approval (unless they're significant and make my review of the code obsolete or something).

Reply:
![Screenshot_20240324_174903_Samsung Internet](https://github.com/ArchipelagoMW/Archipelago/assets/57900059/25c171d0-2d0c-46aa-a7ac-da5ac4509de2)

I don't read commit names either, just showing that yes, it was considered :D